domain/macaroon/state/rootkeys.go

```diff
@@ -0,0 +1,122 @@
+)
+
+// RootKeyState describes the persistence layer for
+// macaroon root keys
+type RootKeyState struct {
+	*domain.StateBase
+}
+
+// NewRootKeyState return a new macaroon root key state reference
+func NewRootKeyState(factory coredatabase.TxnRunnerFactory) *RootKeyState {
+	return &RootKeyState{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetKey gets the key with a given id from dqlite. If not key is found, a
+// macaroonerrors.KeyNotFound error is returned.
+func (st *RootKeyState) GetKey(ctx context.Context, id []byte) (macaroon.RootKey, error) {
+	db, err := st.DB()
+	if err != nil {
+		return macaroon.RootKey{}, errors.Trace(err)
+	}
+	key := macaroon.RootKey{
+		Id: id,
+	}
+	getKeyStmt, err := st.Prepare("SELECT &RootKey.* FROM macaroon_root_key WHERE id = $RootKey.id", key)
+	if err != nil {
+		return macaroon.RootKey{}, errors.Annotate(err, "preparing get root key statement")
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, getKeyStmt, key).Get(&key)
+		if internaldatabase.IsErrNotFound(err) {
+			return errors.Annotatef(macaroonerrors.KeyNotFound, "key with id %s", string(id))
+		}
+		return domain.CoerceError(err)
+	})
+	return key, errors.Trace(err)
+}
+
+// FindLatestKey returns the most recently created root key k following all
+// the conditions:
+//
+// k.Created >= createdAfter
+// k.Expires >= expiresAfter
+// k.Expires <= expiresBefore
+//
+// If no such key was found, return a macaroonerrors.KeyNotFound error
+func (st *RootKeyState) FindLatestKey(ctx context.Context, createdAfter, expiresAfter, expiresBefore time.Time) (macaroon.RootKey, error) {
+	db, err := st.DB()
+	if err != nil {
+		return macaroon.RootKey{}, errors.Trace(err)
+	}
+
+	var key macaroon.RootKey
+	m := sqlair.M{
+		"created_after":  createdAfter,
+		"expires_after":  expiresAfter,
+		"expires_before": expiresBefore,
+	}
+
+	q := `
+SELECT &RootKey.*
+FROM   macaroon_root_key
+WHERE  created_at >= $M.created_after
+AND    expires_at >= $M.expires_after
+AND    expires_at <= $M.expires_before
+ORDER BY created_at DESC
+`
+	stmt, err := st.Prepare(q, key, m)
+	if err != nil {
+		return macaroon.RootKey{}, errors.Annotatef(err, "preparing get latest key statement")
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, m).Get(&key)
+		if database.IsErrNotFound(err) {
+			return macaroonerrors.KeyNotFound
+		}
+		return domain.CoerceError(err)
+	})
+	return key, errors.Trace(err)
+}
+
+// InsertKey inserts the given root key into dqlite. If a key with matching
+// id already exists, return a macaroonerrors.KeyAlreadyExists error.
+func (st *RootKeyState) InsertKey(ctx context.Context, key macaroon.RootKey) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertKeyStmt, err := st.Prepare("INSERT INTO macaroon_root_key (*) VALUES ($RootKey.*)", macaroon.RootKey{})
```

Comment:
You've got the incoming key to prepare with, so you don't need to create a zero-value one here.

Reply:
Personally I do prefer using zero-valued structs in prepare statements.

However, I know this is a conversation which has already been had. So I don't want to dig up old bones

Reply:
Done